File: src/ios/Sources/AnsightToolsFileSystem/DeleteFileTool.swift
import AnsightCore
import Foundation

public final class DeleteFileTool: AnsightTool {
    private let options: AnsightFileSystemToolsOptions

    public init(options: AnsightFileSystemToolsOptions = .default) {
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightFileSystemToolIds.deleteFile,
            name: "Delete File",
            description: "Deletes a sandboxed file from an approved root.",
            category: "files",
            policy: .critical,
            keywords: "filesystem file delete remove sandbox",
            argumentsSchema: AnsightFileSystemToolSchemas.deleteFileArguments,
            resultSchema: AnsightFileSystemToolSchemas.deleteFileResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            let roots = try AnsightFileSystemSandbox.roots(options: options)
            let resolvedFile = try AnsightFileSystemSandbox.resolvePath(
                arguments: arguments,
                roots: roots,
                requireExisting: true,
                expectDirectory: false
            )
            let attributes = try AnsightFileSystemIO.attributes(path: resolvedFile.fullPath)
            guard case .object(var payload) = AnsightFileSystemContentDescriptor.resolvedFilePayload(
                resolvedFile: resolvedFile,
                roots: roots,
                attributes: attributes
            ) else {
                return .failure("Failed to build delete payload.", errorCode: "filesystem_delete_failed")
            }

            try FileManager.default.removeItem(atPath: resolvedFile.fullPath)
            payload["deleted"] = .bool(true)
            payload["capturedAtUtc"] = .string(AnsightClock.isoNow())
            return .success(.object(payload))
        } catch {
            return .failure(error.localizedDescription, errorCode: "filesystem_delete_failed")
        }
    }
}

File: src/dotnet/Ansight.Tools.FileSystem/DeleteFileTool.cs
namespace Ansight.Tools.FileSystem;

public sealed class DeleteFileTool : ITool
{
    private readonly FileSystemToolsOptions options;

    public DeleteFileTool(FileSystemToolsOptions? options = null)
    {
        this.options = options ?? FileSystemToolsOptions.Default;
    }

    public string Category => "files";

    public ToolPolicy Policy => ToolPolicy.Critical;

    public string Id => FileSystemToolIds.DeleteFile;

    public string Name => "Delete File";

    public string Description => "Deletes a file from a sandbox-constrained path.";

    public string Keywords => "filesystem file delete remove sandbox";

    public ToolSchema ArgumentsSchema => FileSystemToolSchemas.DeleteFileArguments;

    public ToolSchema ResultSchema => FileSystemToolSchemas.DeleteFileResult;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            var roots = FileSystemSandbox.GetRoots(options);
            var resolvedFile = FileSystemSandbox.ResolvePath(
                arguments,
                roots,
                requireExisting: true,
                expectDirectory: false);
            var fileInfo = new FileInfo(resolvedFile.FullPath);
            var payload = FileSystemContentDescriptor.CreateResolvedFilePayload(resolvedFile, roots, fileInfo);

            File.Delete(resolvedFile.FullPath);

            payload["deleted"] = true;
            payload["capturedAtUtc"] = DateTime.UtcNow.ToString("O");

            return Task.FromResult(ToolResult.Success(payload));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "filesystem_delete_failed"));
        }
    }
}
